staging/src/k8s.io/apimachinery/pkg/api/validate/enum.go

```diff
@@ -0,0 +1,35 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validate
+
+import (
+	"slices"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+// Enum verifies that the specified value is one of the valid symbols.
+// This is for string enums only.
+func Enum[T ~string](fldPath *field.Path, value T, symbols sets.Set[T]) field.ErrorList {
+	if !symbols.Has(value) {
+		symbolList := symbols.UnsortedList()
```

Comment:
I wonder if this might end up being a perf thing - should we create a variable for the sorted-list as well as the set?  Or maybe we just pass the sorted list here and do a linear or binary search of it, instead of `set.Has()`.  A good TODO for later.

Reply:
I've been working under the assumption that passing validation should be fast, but failing validation is far less common and, can be slower, and should bias toward producing high quality error messages.  The list sort here only happens if the symbol is not in the set.

Reply:
The problem is that failing validation is something any authorized user can do, and if it is slow, it's a potential DOS?  I am sure this alone isn't going to cause a DOS, but it seems easy to avoid.